src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs

```diff
@@ -1242,6 +1242,16 @@ public virtual MethodSymbol TupleUnderlyingMethod
             }
         }
 
+        /// <summary>
+        /// Returns a suitable location for this MethodSymbol for a diagnostic to be placed at. This
+        /// is generally just the first entry in <see cref="Symbol.Locations"/>, but it may be a 
+        /// smaller location if the full location is large.  For example, Locations[0] on  lambda 
+        /// symbols covers the entire syntax, which is inconvenient but remains for compatibility.
+        /// For better diagnostics quality, <see cref="DiagnosticLocation"/> will point to the 
+        /// "delegate" or the "=>".
+        /// </summary>
+        internal virtual Location DiagnosticLocation => Locations[0];
```

Comment:
Before calling `Locations[0]`, a caller could reasonably check `Locations.IsEmpty`. Before calling `DiagnosticLocation`, it's not clear what to check or even that a check is necessary.

As to the particular approach for this bug: casting to `LambdaSymbol` in this one caller is probably the simplest change  and would avoid any new property that could be misused. An alternative would be to have `MethodSymbol.DiagnosticLocation` return `Locations.IsEmpty ? NoLocation : Locations[0]`. The cast seems preferable to me.